primitives/src/circuit/merkle_tree/sparse_merkle_tree.rs

```diff
@@ -0,0 +1,581 @@
+        )?;
+        self.enforce_true(bool_val.into())
+    }
+}
+
+trait SparseMerkleTreeHelperGadget<F: RescueParameter> {
+    /// Computes the merkle root based on an empty leaf element and a
+    /// merkle path.
+    /// * `path_vars` - variables corresponding to the Merkle path.
+    /// * `return` - variable corresponding to the root value of the Merkle
+    ///   tree.
+    fn compute_merkle_root_for_empty(
+        &mut self,
+        path_vars: &Merkle3AryNonMembershipProofVar,
+    ) -> Result<Variable, CircuitError>;
+}
+
+impl<F: RescueParameter> SparseMerkleTreeHelperGadget<F> for PlonkCircuit<F> {
+    fn compute_merkle_root_for_empty(
+        &mut self,
+        path_vars: &Merkle3AryNonMembershipProofVar,
+    ) -> Result<Variable, CircuitError> {
+        let mut cur_label = self.zero();
+        for cur_node in path_vars.node_vars.iter() {
+            let input_labels =
+                MerkleTreeHelperGadget::<SparseMerkleTree<F>>::constrain_sibling_order(
+                    self,
+                    cur_label,
+                    cur_node.sibling1,
+                    cur_node.sibling2,
+                    cur_node.is_left_child,
+                    cur_node.is_right_child,
+                )?;
+            // check that the left child's label is non-zero
+            self.non_zero_gate(input_labels[0])?;
+            cur_label =
+                RescueNativeGadget::<F>::rescue_sponge_no_padding(self, &input_labels, 1)?[0];
+        }
+        Ok(cur_label)
+    }
+}
+
+impl<F: RescueParameter> MerkleTreeHelperGadget<SparseMerkleTree<F>> for PlonkCircuit<F> {
+    type MembershipProofVar = Merkle3AryMembershipProofVar;
+
+    fn constrain_sibling_order(
+        &mut self,
+        node: Variable,
+        sib1: Variable,
+        sib2: Variable,
+        node_is_left: BoolVar,
+        node_is_right: BoolVar,
+    ) -> Result<[Variable; 3], CircuitError> {
+        let one = F::one();
+        let left_node = self.conditional_select(node_is_left, sib1, node)?;
+        let right_node = self.conditional_select(node_is_right, sib2, node)?;
+        let left_plus_right = self.add(left_node, right_node)?;
+        let mid_node = self.lc(
+            &[node, sib1, sib2, left_plus_right],
+            &[one, one, one, one.neg()],
+        )?;
+        Ok([left_node, mid_node, right_node])
+    }
+
+    fn constrain_membership_proof(
+        &mut self,
+        merkle_proof: &MembershipProof<F>,
+    ) -> Result<Self::MembershipProofVar, CircuitError> {
+        let path = <BigUint as ToTraversalPath<U3>>::to_traversal_path(
+            &merkle_proof.pos,
+            merkle_proof.tree_height() - 1,
+        );
+
+        let leaf_elem = match merkle_proof.elem() {
+            Some(elem) => elem,
+            None => {
+                return Err(CircuitError::InternalError(
+                    "The proof doesn't contain a leaf element".to_string(),
+                ))
+            },
```

Comment:
should this be `ParameterError` instead? (conventionally, we use `InternalError` to represents something wrong with our code whereas this is a user-provided input which might be malformed)

Reply:
You're absolutely right.

Reply:
Done in c0acab04c0d94c9df4c38a04927bcefa0f2f6f2d